File: src/pablo/doctor.py
"""CLI preflight checks — ``pablo doctor`` / ``/pablo-doctor``.

A Python module (decided 2026-07-26): derives the required CLI set from the
configured projects. ``gh`` is always required (every project's PR/CI flow
goes through GitHub); the Atlassian CLI ``acli`` (Jira) — and a separate
``acli-confluence`` probe when some project configures a Confluence space —
only for jira projects; ``linear`` only when a project uses that provider;
``orca``/``opencode`` always (the agent-runner path). Each CLI is checked
for being installed AND authenticated/ready, surfacing the CLI's own
instructions on failure. Non-zero exit when anything required is missing —
the dispatcher uses the same check as its fail-fast guard.
"""

from __future__ import annotations

import shutil
import subprocess
from dataclasses import dataclass

from pablo.config import ProjectConfig

# cli name → (probe argv, login hint shown on failure). The probe argv's
# first element is also the binary name used for the PATH check.
CLI_PROBES: dict[str, tuple[list[str], str]] = {
    "gh": (["gh", "auth", "status"], "run: gh auth login"),
    "acli": (["acli", "jira", "auth", "status"], "run: acli auth login"),
    "acli-confluence": (
        ["acli", "confluence", "auth", "status"],
        "run: acli confluence auth login",
    ),
    "linear": (["linear", "auth", "status"], "run: linear auth login (schpet/linear-cli)"),
    "orca": (["orca", "status"], "start the Orca app, or run: orca open"),
    "opencode": (["opencode", "--version"], "install opencode: https://opencode.ai"),
}

ALWAYS_REQUIRED = ["gh", "opencode", "orca"]


@dataclass
class CheckResult:
    cli: str
    installed: bool
    authenticated: bool
    detail: str
    hint: str

    @property
    def ok(self) -> bool:
        return self.installed and self.authenticated


def required_clis(projects: dict[str, ProjectConfig]) -> list[str]:
    required = set(ALWAYS_REQUIRED)
    needs_confluence = False
    for cfg in projects.values():
        if cfg.provider == "jira":
            required.add("acli")
            if cfg.confluence_space:
                needs_confluence = True
        elif cfg.provider == "linear":
            required.add("linear")
    if needs_confluence:
        required.add("acli-confluence")
    return sorted(required)


PROBE_TIMEOUT_S = 30


def _probe(argv: list[str]) -> tuple[int, str]:
    try:
        proc = subprocess.run(
            argv,
            capture_output=True,
            text=True,
            timeout=PROBE_TIMEOUT_S,
            stdin=subprocess.DEVNULL,
        )
    except subprocess.TimeoutExpired:
        return 124, f"timed out after {PROBE_TIMEOUT_S}s"
    output = (proc.stderr.strip() + "\n" + proc.stdout.strip()).strip()
    return proc.returncode, output


def _binary_for(cli_name: str) -> str:
    """The PATH binary to look up for a check — the probe's first argv
    element is always the binary name (acli appears in two probes)."""
    return CLI_PROBES[cli_name][0][0]


def check_all(projects: dict[str, ProjectConfig]) -> list[CheckResult]:
    results = []
    for cli_name in required_clis(projects):
        probe_argv, hint = CLI_PROBES[cli_name]
        if shutil.which(_binary_for(cli_name)) is None:
            results.append(
                CheckResult(
                    cli=cli_name,
                    installed=False,
                    authenticated=False,
                    detail=f"{cli_name} is not installed (not on PATH)",
                    hint=hint,
                )
            )
            continue
        code, output = _probe(probe_argv)
        results.append(
            CheckResult(
                cli=cli_name,
                installed=True,
                authenticated=code == 0,
                detail=output.splitlines()[0] if output else ("ok" if code == 0 else "failed"),
                hint=hint if code != 0 else "",
            )
        )
    return results


def render(results: list[CheckResult]) -> str:
    lines = []
    for result in results:
        icon = "✅" if result.ok else "❌"
        lines.append(f"{icon} {result.cli:<10} {result.detail}")
        if not result.ok and result.hint:
            lines.append(f"   → {result.hint}")
    return "\n".join(lines)

File: app/src/Doctor/Doctor.php
<?php

declare(strict_types=1);

namespace Pablo\Doctor;

use Pablo\Agents\AgentLauncherFactory;
use Pablo\Config\ProjectConfig;
use Pablo\Support\ProcessRunnerInterface;
use Symfony\Component\Process\ExecutableFinder;

/**
 * CLI preflight checks — pablo system:doctor / /pablo-doctor.
 *
 * Derives the required CLI set from the configured projects. gh is always
 * required; acli (Jira) and a separate acli-confluence probe when a project
 * configures a Confluence space, only for jira projects; linear only when a
 * project uses that provider; orca/opencode always; openchamber only when the
 * global agent_backend selects it.
 */
final class Doctor
{
    private const CLI_PROBES = [
        'gh' => [['gh', 'auth', 'status'], 'run: gh auth login'],
        'acli' => [['acli', 'jira', 'auth', 'status'], 'run: acli auth login'],
        'acli-confluence' => [['acli', 'confluence', 'auth', 'status'], 'run: acli confluence auth login'],
        'linear' => [['linear', 'auth', 'status'], 'run: linear auth login (schpet/linear-cli)'],
        'orca' => [['orca', 'status'], 'start the Orca app, or run: orca open'],
        'opencode' => [['opencode', '--version'], 'install opencode: https://opencode.ai'],
        'openchamber' => [['openchamber', 'status'], 'install openchamber: https://github.com/openchamber/openchamber'],
    ];

    private const ALWAYS_REQUIRED = ['gh', 'opencode', 'orca'];

    /**
     * Wizard ordering: hard requirements first, then the per-provider
     * trackers (only included when a project uses them).
     */
    public const CLI_ORDER = ['gh', 'orca', 'opencode', 'openchamber', 'acli', 'acli-confluence', 'linear'];

    /** @var array<string, array<string, string>> platform => install command */
    private const INSTALL_COMMANDS = [
        'gh' => [
            'Darwin' => 'brew install gh',
            'Linux' => 'sudo apt-get install gh',
        ],
        'opencode' => [
            'Darwin' => 'curl -fsSL https://opencode.ai/install | bash',
            'Linux' => 'curl -fsSL https://opencode.ai/install | bash',
        ],
        'openchamber' => [
            'Darwin' => 'curl -fsSL https://raw.githubusercontent.com/openchamber/openchamber/main/scripts/install.sh | bash',
            'Linux' => 'curl -fsSL https://raw.githubusercontent.com/openchamber/openchamber/main/scripts/install.sh | bash',
        ],
        'acli' => [
            'Darwin' => 'brew tap atlassian-labs/acli && brew install acli',
            'Linux' => 'brew tap atlassian-labs/acli && brew install acli',
        ],
        'acli-confluence' => [
            'Darwin' => 'brew tap atlassian-labs/acli && brew install acli',
            'Linux' => 'brew tap atlassian-labs/acli && brew install acli',
        ],
        'linear' => [
            'Darwin' => 'brew install schpet/tap/linear',
            'Linux' => 'deno install -A --reload -f -g -n linear jsr:@schpet/linear-cli',
        ],
    ];

    private const DOCS_URLS = [
        'gh' => 'https://cli.github.com',
        'opencode' => 'https://opencode.ai',
        'openchamber' => 'https://github.com/openchamber/openchamber',
        'acli' => 'https://developer.atlassian.com/cloud/acli/',
        'acli-confluence' => 'https://developer.atlassian.com/cloud/acli/',
        'linear' => 'https://github.com/schpet/linear-cli',
    ];

    private const PROBE_TIMEOUT_S = 30;

    /** @var callable|null PATH-lookup override fn(string): ?string */
    private $which;

    public function __construct(
        private readonly ProcessRunnerInterface $runner,
        private readonly AgentLauncherFactory $agentLaunchers,
        ?callable $which = null,
    ) {
        $this->which = $which;
    }

    /**
     * @param array<string, ProjectConfig> $projects
     *
     * @return list<string>
     */
    public function requiredClis(array $projects): array
    {
        $required = array_fill_keys(self::ALWAYS_REQUIRED, true);
        $needsConfluence = false;
        foreach ($projects as $cfg) {
            if ('jira' === $cfg->provider) {
                $required['acli'] = true;
                if (null !== $cfg->confluenceSpace) {
                    $needsConfluence = true;
                }
            } elseif ('linear' === $cfg->provider) {
                $required['linear'] = true;
            }
        }
        if ($needsConfluence) {
            $required['acli-confluence'] = true;
        }
        if ('openchamber' === $this->agentLaunchers->resolveBackend()) {
            $required['openchamber'] = true;
        }
        $list = array_keys($required);
        sort($list);

        return $list;
    }

    /** @return list<string> */
    public function cliOrder(): array
    {
        return self::CLI_ORDER;
    }

    public function installCommand(string $cli): ?string
    {
        $perOs = self::INSTALL_COMMANDS[$cli] ?? null;
        if (null === $perOs) {
            return null;
        }

        return $perOs[\PHP_OS_FAMILY] ?? null;
    }

    public function docsUrl(string $cli): ?string
    {
        return self::DOCS_URLS[$cli] ?? null;
    }

    private function binaryFor(string $cliName): string
    {
        return self::CLI_PROBES[$cliName][0][0];
    }

    private function which(string $bin): ?string
    {
        if (null !== $this->which) {
            return ($this->which)($bin);
        }

        return (new ExecutableFinder())->find($bin);
    }

    /**
     * @param array<string, ProjectConfig> $projects
     *
     * @return list<CheckResult>
     */
    public function checkAll(array $projects): array
    {
        return $this->checkIds($this->requiredClis($projects));
    }

    /**
     * @param list<string> $cliIds
     *
     * @return list<CheckResult>
     */
    public function checkIds(array $cliIds): array
    {
        $results = [];
        foreach ($cliIds as $cliName) {
            [$probeArgv, $hint] = self::CLI_PROBES[$cliName];
            if (null === $this->which(self::binaryFor($cliName))) {
                $results[] = new CheckResult(
                    cli: $cliName,
                    installed: false,
                    authenticated: false,
                    detail: "{$cliName} is not installed (not on PATH)",
                    hint: $hint,
                );
                continue;
            }
            $probe = $this->runner->probe($probeArgv, self::PROBE_TIMEOUT_S);
            $lines = preg_split('/\r?\n/', $probe->output) ?: [];
            $results[] = new CheckResult(
                cli: $cliName,
                installed: true,
                authenticated: 0 === $probe->exitCode,
                detail: '' !== $probe->output ? ($lines[0] ?? '') : (0 === $probe->exitCode ? 'ok' : 'failed'),
                hint: 0 !== $probe->exitCode ? $hint : '',
            );
        }

        return $results;
    }

    /** @param array<int, CheckResult> $results */
    public function render(array $results): string
    {
        $lines = [];
        foreach ($results as $result) {
            $icon = $result->ok() ? '✅' : '❌';
            $lines[] = \sprintf("{$icon} %-10s %s", $result->cli, $result->detail);
            if (!$result->ok() && '' !== $result->hint) {
                $lines[] = "   → {$result->hint}";
            }
        }

        return implode("\n", $lines);
    }
}
